kernel/libs/aster-bigtcp/src/socket/bound.rs

```diff
@@ -99,146 +129,172 @@ impl RawTcpSocketExt {
-    fn set_dead_timewait(&self, socket: &RawTcpSocketExt) {
-        debug_assert!(
-            socket.in_background && socket.state() == smoltcp::socket::tcp::State::TimeWait
-        );
+    fn set_dead_timewait(&self, socket: &RawTcpSocketExt<E>) {
+        debug_assert!(socket.state() == smoltcp::socket::tcp::State::TimeWait);
         self.is_dead.store(true, Ordering::Relaxed);
     }
 }
 
-impl<E: Ext> AnySocket<E> for TcpSocket {
-    type RawSocket = RawTcpSocket;
+impl<E: Ext> Inner<E> for TcpConnectionInner<E> {
     type Observer = E::TcpEventObserver;
 
-    fn new(socket: Box<Self::RawSocket>) -> Self {
-        let socket_ext = RawTcpSocketExt {
-            socket,
-            has_connected: false,
-            in_background: false,
-        };
-
-        Self {
-            socket: SpinLock::new(socket_ext),
-            is_dead: AtomicBool::new(false),
-        }
-    }
-
-    fn on_drop(this: &Arc<BoundSocketInner<Self, E>>) {
-        let mut socket = this.socket.lock();
+    fn on_drop(this: &KeyableArc<SocketBg<Self, E>>) {
+        let mut socket = this.inner.lock();
 
-        socket.in_background = true;
         socket.close();
 
-        // A TCP socket may not be appropriate for immediate removal. We leave the removal decision
-        // to the polling logic.
+        // A TCP connection may not be appropriate for immediate removal. We leave the removal
+        // decision to the polling logic.
         this.update_next_poll_at_ms(PollAt::Now);
-        this.socket.update_dead(&socket);
+        TcpConnectionInner::update_dead(this, &socket);
     }
 }
 
-/// States needed by [`BoundUdpSocketInner`] but not [`BoundTcpSocketInner`].
-type UdpSocket = SpinLock<Box<RawUdpSocket>, LocalIrqDisabled>;
+pub struct TcpBacklog<E: Ext> {
+    socket: Box<RawTcpSocket>,
+    max_conn: usize,
+    connecting: BTreeSet<TcpConnection<E>>,
+    connected: Vec<TcpConnection<E>>,
+}
 
-impl<E: Ext> AnySocket<E> for UdpSocket {
-    type RawSocket = RawUdpSocket;
-    type Observer = E::UdpEventObserver;
+pub type TcpListenerInner<E> = SpinLock<TcpBacklog<E>, LocalIrqDisabled>;
+
+impl<E: Ext> Inner<E> for TcpListenerInner<E> {
+    type Observer = E::TcpEventObserver;
 
-    fn new(socket: Box<Self::RawSocket>) -> Self {
-        Self::new(socket)
+    fn on_drop(this: &KeyableArc<SocketBg<Self, E>>) {
+        // A TCP listener can be removed immediately.
+        this.bound.iface().common().remove_tcp_listener(this);
+
+        let (connecting, connected) = {
+            let mut socket = this.inner.lock();
+            (
+                core::mem::take(&mut socket.connecting),
+                core::mem::take(&mut socket.connected),
+            )
+        };
+
+        // The lock on `connecting`/`connected` cannot be locked after locking `self`, otherwise we
+        // might get a deadlock. due to inconsistent lock order problems.
+        drop(connecting);
+        drop(connected);
```

Comment:
Should we send RST to remote side for backlog connections? Otherwise how can remote side know the listener is dropped?

Reply:
The current implementation goes through the normal socket close process, which sends FIN packets to the remote end so that the remote end _can_ know that the listener is closed. This was our behavior prior to this PR.

However, I checked the Linux behavior. It will send RST packets when the listener closes.
```text
> python
Python 3.12.7 (main, Oct  1 2024, 11:15:50) [GCC 14.2.1 20240910] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import socket
>>> s = socket.socket()
>>> s.listen()
>>> s
<socket.socket fd=3, family=2, type=1, proto=0, laddr=('0.0.0.0', 46913)>
>>> t = socket.socket()
>>> t.connect(('127.0.0.1', 46913))
>>> s.close()
>>> t.recv(1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ConnectionResetError: [Errno 104] Connection reset by peer
```

I'll add a FIXME here and leave this for future work, since we don't currently have good support for connection resets anyway. A few things are missing, so it's hard to write reasonable regression tests:
 - We don't support `POLLERR`, and we don't even record whether the connection is reset.
 - Also, `send`/`recv` can never report `ECONNRESET`.
 - We don't support `SO_LINGER`, so it's hard to manually trigger a connection reset.

Reply:
It's fine to leave it as future work and just add a README now.